vpr/src/place/analytical_placement/AnalyticalSolver.cpp

```diff
@@ -162,21 +211,255 @@ void QPHybridSolver::solve(unsigned iteration, PartialPlacement &p_placement) {
+    //     double pseudo_w = coeff_pseudo_anchor; 
+    //     A_sparse_x.coeffRef(i, i) += pseudo_w;
+    //     A_sparse_y.coeffRef(i, i) += pseudo_w;
+    //     b_x(i) += pseudo_w * node_loc_x_legalized[i];
+    //     b_y(i) += pseudo_w *  node_loc_x_legalized[i];
+    // }
+}
+
+// This function is the inner loop that alternate between building matrix based on placement, and solving model to get placement.
+void B2BSolver::b2b_solve_loop(unsigned iteration, PartialPlacement &p_placement){
+    double previous_hpwl, current_hpwl = std::numeric_limits<double>::max();
+    unsigned counter = 0;
+    do{
+        counter++; 
+        previous_hpwl = current_hpwl;
+        VTR_LOG("placement hpwl in b2b loop: %f\n", p_placement.get_HPWL());
+        VTR_ASSERT(p_placement.is_valid_partial_placement() && "did not produce a valid placement in b2b solve loop");
+        populate_matrix(p_placement);
+        if(iteration != 0)
+            populate_matrix_anchor(p_placement, iteration);
+        Eigen::VectorXd x, y;
+        Eigen::ConjugateGradient<Eigen::SparseMatrix<double>, Eigen::Lower|Eigen::Upper> cg_x;
+        Eigen::ConjugateGradient<Eigen::SparseMatrix<double>, Eigen::Lower|Eigen::Upper> cg_y;
+        VTR_ASSERT(!b_x.hasNaN() && "b_x has NaN!");
+        VTR_ASSERT(!b_y.hasNaN() && "b_y has NaN!");
+        VTR_ASSERT((b_x.array() >= 0).all() && "b_x has NaN!");
+        VTR_ASSERT((b_y.array() >= 0).all() && "b_y has NaN!");
+        cg_x.compute(A_sparse_x);
+        cg_y.compute(A_sparse_y);
+        // Not worth the time
+        // having more cg iteration does not help with over convergence
+        // cg_x.setMaxIterations(cg_x.maxIterations() * 100);
+        // cg_x.setTolerance(cg_x.tolerance() * 100);
+        // cg_y.setMaxIterations(cg_y.maxIterations() * 100);
+        // cg_y.setTolerance(cg_y.tolerance() * 100);
+        VTR_ASSERT(cg_x.info() == Eigen::Success && "Conjugate Gradient failed at compute for A_x!");
+        VTR_ASSERT(cg_y.info() == Eigen::Success && "Conjugate Gradient failed at compute for A_y!");
+        x = cg_x.solve(b_x);
+        y = cg_y.solve(b_y);
+        // VTR_ASSERT(cg_x.info() == Eigen::Success && "Conjugate Gradient failed at solving b_x!");
+        // VTR_ASSERT(cg_y.info() == Eigen::Success && "Conjugate Gradient failed at solving b_y!");
+        for (size_t node_id = 0; node_id < p_placement.num_moveable_nodes; node_id++) {
+            p_placement.node_loc_x[node_id] = x[node_id];
+            p_placement.node_loc_y[node_id] = y[node_id];
+        }
+        current_hpwl = p_placement.get_HPWL();
+    // This would not work because sometimes they stay in close proximity before growing or shrinking
+    // }while(std::abs(current_hpwl - previous_hpwl) < 5 && counter < 100);
+    // current_hpwl > previous_hpwl - 10 would not work when this tries to grow and converge to legalized solution
+    // simplest one for now
+    }while(counter < 30);
```

Comment:
Spacing. Also I would make 30 a member of the class so it can be changed easily (similar to my comment on epsilon).